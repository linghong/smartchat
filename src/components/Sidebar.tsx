import React, { FC } from 'react'

import MenuItem from './MenuItem'

interface SidebarProps {
  setIsSidebarOpen: (isSidebarOpen: boolean) => void;
  messageSubjectList: string[]
}

const Sidebar : FC<SidebarProps>= ({ setIsSidebarOpen, messageSubjectList }) => {

  return (
    <div className="bg-slate-500 text-slate-50 h-full w-full xs:w-60 sm:w-80 md:w-72 lg:w-56 xl:w-56">
       <ul className="px-2  py-5 ">
          <MenuItem
            key="embedragfile"
            title="Embed RAG File" 
            link="/embedragfile" 
            itemList={["File 1", "File 2", "File 3"]}
            setIsSidebarOpen={setIsSidebarOpen}
          />
          <MenuItem 
            key="finetunemodel" 
            title="Finetune AI Model" 
            link="/finetunemodel" 
            itemList={["My Model1", "My Model2", "My Model3"]}
            setIsSidebarOpen={setIsSidebarOpen}
          />
          <MenuItem 
            key="chatwithai"
            title="Chat With AI" 
            link="/" 
            itemList={[messageSubjectList[0]]}
            setIsSidebarOpen={setIsSidebarOpen}
            defaultOpen={true}
          />
      </ul>
    </div>
  )
}

export default Sidebar
