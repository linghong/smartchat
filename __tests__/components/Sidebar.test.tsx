import React from 'react';
import {
  render,
  screen,
  waitFor,
  act,
  fireEvent
} from '@testing-library/react';
import '@testing-library/jest-dom';

import Sidebar from '@/src/components/Sidebar';
import {
  fetchChats,
  fetchChatMessages,
  deleteChat,
  editChatTitle
} from '@/src/utils/sqliteApiClient';

// Mock the MenuItem component
jest.mock('@/src/components/MenuItem', () => {
  return function MockMenuItem({
    title,
    itemList,
    link,
    defaultOpen,
    onItemClick,
    activeItemId,
    onDeleteClick,
    onEditClick,
    setIsSidebarOpen
  }: {
    title: string;
    itemList: any[];
    link: string;
    defaultOpen: boolean;
    onItemClick?: (id: string) => void;
    activeItemId?: string;
    onDeleteClick?: (id: string) => void;
    onEditClick?: (id: string) => void;
    setIsSidebarOpen?: (isOpen: boolean) => void;
  }) {
    return (
      <div>
        <a href={link}>
          <h2>{title}</h2>
        </a>
        <ul>
          {defaultOpen &&
            itemList?.map(item => (
              <li
                key={item.value}
                onClick={() => {
                  onItemClick && onItemClick(item.value);
                  setIsSidebarOpen && setIsSidebarOpen(false);
                }}
                data-active={item.value === activeItemId}
              >
                {item.label}
                {onDeleteClick && (
                  <button onClick={() => onDeleteClick(item.value)}>
                    Delete
                  </button>
                )}
                {onEditClick && (
                  <button onClick={() => onEditClick(item.value)}>Edit</button>
                )}
              </li>
            ))}
        </ul>
      </div>
    );
  };
});

// Mock the AIHub component
jest.mock('@/src/components/AIHub', () => {
  return function MockAIHub() {
    return <div>AI Hub</div>;
  };
});

jest.mock('@/src/utils/sqliteApiClient', () => ({
  fetchChats: jest.fn(),
  fetchChatMessages: jest.fn(),
  deleteChat: jest.fn(),
  editChatTitle: jest.fn()
}));

jest.mock('next/router', () => ({
  useRouter: () => ({
    pathname: '/'
  })
}));

const mockSetIsConfigPanelVisible = jest.fn();
const mockSetIsSidebarOpen = jest.fn();
const mockSetChatId = jest.fn();
const mockSetChats = jest.fn();
const mockSetChatHistory = jest.fn();
const mockSetImageSrcHistory = jest.fn();

const mockNamespacesList = [
  { label: 'Namespace 1', value: 'ns1' },
  { label: 'Namespace 2', value: 'ns2' }
];

const mockChats = [
  { value: '1', label: 'Test Chat 1' },
  { value: '2', label: 'Test Chat 2' }
];

describe('Sidebar Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (fetchChats as jest.Mock).mockResolvedValue(mockChats);
    (fetchChatMessages as jest.Mock).mockResolvedValue([
      {
        userMessage: 'Hello',
        aiMessage: 'Hi there',
        model: 'model1',
        images: []
      }
    ]);
  });

  const renderSidebar = (chatId = '0') =>
    render(
      <Sidebar
        isSidebarOpen={true}
        setIsConfigPanelVisible={mockSetIsConfigPanelVisible}
        setIsSidebarOpen={mockSetIsSidebarOpen}
        namespacesList={mockNamespacesList}
        chatId={chatId}
        setChatId={mockSetChatId}
        chats={mockChats}
        setChats={mockSetChats}
        setChatHistory={mockSetChatHistory}
        setImageSrcHistory={mockSetImageSrcHistory}
      />
    );

  it('renders Sidebar component with all expected elements', async () => {
    await act(async () => {
      renderSidebar();
    });

    expect(screen.getByText('Embed RAG File')).toBeInTheDocument();
    expect(screen.getByText('Finetune AI Model')).toBeInTheDocument();
    expect(screen.getByText('Chat With AI')).toBeInTheDocument();
    expect(screen.getByText('AI Hub')).toBeInTheDocument();
  });

  it('handles fetch error gracefully', async () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    (fetchChats as jest.Mock).mockRejectedValue(new Error('Fetch failed'));

    await act(async () => {
      renderSidebar();
    });

    await waitFor(() => {
      expect(consoleSpy).toHaveBeenCalledWith(
        'Failed to fetch chats: Fetch failed'
      );
    });

    consoleSpy.mockRestore();
  });

  it('fetches and renders chats on mount', async () => {
    await act(async () => {
      renderSidebar();
    });

    expect(fetchChats).toHaveBeenCalled();

    await waitFor(() => {
      expect(screen.getByText('Test Chat 1')).toBeInTheDocument();
      expect(screen.getByText('Test Chat 2')).toBeInTheDocument();
    });
  });

  it('handles chat click correctly', async () => {
    await act(async () => {
      renderSidebar();
    });

    await waitFor(() => {
      expect(screen.getByText('Test Chat 1')).toBeInTheDocument();
    });

    const chatItem = screen.getByText('Test Chat 1');
    await act(async () => {
      fireEvent.click(chatItem);
    });

    expect(fetchChatMessages).toHaveBeenCalledWith(1);
    expect(mockSetIsConfigPanelVisible).toHaveBeenCalledWith(false);
    expect(mockSetChatHistory).toHaveBeenCalled();
    expect(mockSetImageSrcHistory).toHaveBeenCalled();
    expect(mockSetChatId).toHaveBeenCalledWith('1');
    expect(mockSetIsSidebarOpen).toHaveBeenCalledWith(false);
  });

  it('handles chat deletion correctly', async () => {
    await act(async () => {
      renderSidebar('1');
    });

    const deleteButton = screen.getAllByText('Delete')[0];
    await act(async () => {
      fireEvent.click(deleteButton);
    });

    expect(deleteChat).toHaveBeenCalledWith('1');
    expect(mockSetChats).toHaveBeenCalled();
    expect(mockSetChatId).toHaveBeenCalledWith('');
    expect(mockSetChatHistory).toHaveBeenCalled();
    expect(mockSetImageSrcHistory).toHaveBeenCalled();
  });

  it('handles chat editing correctly', async () => {
    window.prompt = jest.fn().mockReturnValue('New Chat Title');

    await act(async () => {
      renderSidebar();
    });

    const editButton = screen.getAllByText('Edit')[0];
    await act(async () => {
      fireEvent.click(editButton);
    });

    expect(editChatTitle).toHaveBeenCalledWith('1', 'New Chat Title');
    expect(mockSetChats).toHaveBeenCalled();
  });

  it('handles empty chat messages correctly', async () => {
    (fetchChatMessages as jest.Mock).mockResolvedValue([]);

    await act(async () => {
      renderSidebar();
    });

    const chatItem = await screen.findByText('Test Chat 1');
    await act(async () => {
      fireEvent.click(chatItem);
    });

    expect(mockSetChatHistory).toHaveBeenCalledWith([
      {
        question: '',
        answer: 'Hi, how can I assist you?',
        model: 'Gemini-1.5 Pro Exp'
      }
    ]);
    expect(mockSetImageSrcHistory).toHaveBeenCalledWith([[]]);
  });

  it('handles chat messages with images correctly', async () => {
    (fetchChatMessages as jest.Mock).mockResolvedValue([
      {
        userMessage: 'Hello',
        aiMessage: 'Hi there',
        model: 'model1',
        images: [{ imageFile: 'image1.jpg' }, { imageFile: 'image2.jpg' }]
      }
    ]);

    await act(async () => {
      renderSidebar();
    });

    const chatItem = await screen.findByText('Test Chat 1');
    await act(async () => {
      fireEvent.click(chatItem);
    });

    expect(mockSetImageSrcHistory).toHaveBeenCalledWith([
      ['image1.jpg', 'image2.jpg']
    ]);
  });

  it('matches snapshot', async () => {
    let asFragment: (() => DocumentFragment) | undefined;

    await act(async () => {
      const result = renderSidebar();
      asFragment = result.asFragment;
    });

    if (asFragment) {
      expect(asFragment()).toMatchSnapshot();
    } else {
      throw new Error('asFragment was not assigned');
    }
  });
});
