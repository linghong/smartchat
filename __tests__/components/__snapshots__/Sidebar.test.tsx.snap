// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component Sidebar component matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="px-2 py-5 "
  >
    <li
      class="mt-5 my-8 font-semibold"
    >
      <div
        class="flex justify-between items-center px-3 py-1 border-b hover:bg-slate-500 focus:bg-indigo-100 cursor-pointer text-slate-50"
      >
        <a
          class="flex-grow px-2 hover:underline "
          href="/embedragfile"
        >
          Embed RAG File
        </a>
        <span
          class="cursor-pointer"
          role="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </div>
    </li>
    <li
      class="mt-5 my-8 font-semibold"
    >
      <div
        class="flex justify-between items-center px-3 py-1 border-b hover:bg-slate-500 focus:bg-indigo-100 cursor-pointer text-slate-50"
      >
        <a
          class="flex-grow px-2 hover:underline "
          href="/finetunemodel"
        >
          Finetune AI Model
        </a>
        <span
          class="cursor-pointer"
          role="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </div>
    </li>
    <li
      class="mt-5 my-8 font-semibold"
    >
      <div
        class="flex justify-between items-center px-3 py-1 border-b hover:bg-slate-500 focus:bg-indigo-100 cursor-pointer text-slate-50"
      >
        <a
          class="flex-grow px-2 hover:underline "
          href="/"
        >
          Chat With AI
        </a>
        <span
          class="cursor-pointer"
          role="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
            />
          </svg>
        </span>
      </div>
      <ul
        class="px-6 py-2 font-medium text-slate-200"
      >
        <li
          class="px-2 py-2 hover:bg-slate-400 focus:bg-indigo-100"
        >
          Test Subject
        </li>
      </ul>
    </li>
  </ul>
</DocumentFragment>
`;
